### PR TITLE
Disable JetBrains local ports forwarding with FF

### PR DESCRIPTION
**Pull request by mustard-mh**
## Related Issue(s)

Fixes ENT-826

## How to test

- Try to enable or disable [FeatureFlag](https://app.configcat.com/08da1258-64fb-4a8e-8a1e-51de773884f6/08da1258-6541-4fc7-8b61-c8b47f82f3a0/08da1258-6512-4ec0-80a3-3f6aa301f853?settingId=567437)
- Restart or create workspace with IntelliJ IDEA (latest or stable, but not pinned) after update FF (needs server aware FF changes, i.e. after 5 minutes)
- Check ports view, ports should not expose to local (or access `localhost:<port>`)

|Disabled Ports Exposing|Enabled|
|:-:|:-:|
|SCR-20241009-opju|SCR-20241009-oros|

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-jb-pf</li>
	<li><b>🔗 URL</b> - <a href="https://hw-jb-pf.preview.gitpod-dev.com/workspaces" target="_blank">hw-jb-pf.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-jb-pf-gha.29393</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-jb-pf%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-dev-preview" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by mustard-mh** [comment]
@iQQBot could you have a look if you have time?

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt [hunk 1]
```diff
@@ -100,7 +99,15 @@ abstract class AbstractGitpodPortForwardingService : GitpodPortForwardingService
         return completableFuture
     }
 
+    private fun isPortExposingDisabled(): Boolean {
+        return System.getenv("GITPOD_DISABLE_JETBRAINS_LOCAL_PORT_EXPOSE")?.toBoolean() ?: false
+    }
+
     private fun syncPortsListWithClient(response: Status.PortsStatusResponse) {
+        if (isPortExposingDisabled()) {
```
Comment (akosyakov) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt:107]: @mustard-mh I'm trying to understand the goal of the PR, could you help me please? 🙏 

We have local forwarded and exposed on Interent ports? For local forwarded ports some are forwarded by us other by JB itself, for instance via run configurations and native terminals. My undertand the goal is that:
- we need to disable forwarded ports regardless whether it is from us or JB? do we cover JB case?
- we don't need to disable exposed ports? Does this PR disable them as well? The view should not be completely empty right?

Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt:107]: > we need to disable forwarded ports regardless whether it is from us or JB? do we cover JB case?

@akosyakov Thanks for you review. Yes we do cover them all, I tested with current implements, JetBrains ports are covered as well, see screenshots below

|Enable|Disable|
|:-:|:-:|
|image|image|


> we don't need to disable exposed ports? Does this PR disable them as well?

Exposed ports are exposed by supervisor / ws-proxy , but not JetBrains IDEs. From JetBrains' UI, if you click the add button on the ports view, options all looks like port forwarding (`local -> remote`, or `remote -> local`)

So this view is not a good place for us to show exposed ports

I will write (prepare) a draft message for sales to communicate with users -> they may needs to use in workspace GP-CLI  i.e.`gp ports list` to list and process with ports after this changes

image

Comment (mustard-mh) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt:107]: Btw, the naming is confusing, I'm going to change env name and function name from `local exposing` to `local forwarding`

Comment (akosyakov) [reply components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt]: > Exposed ports are exposed by supervisor / ws-proxy , but not JetBrains IDEs. From JetBrains' UI, if you click the add button on the ports view, options all looks like port forwarding (local -> remote, or remote -> local)

It is true, but the backend plugin monitors add adds each port like local forwarded and exposed. We should keep exposed, only remove local forwarded.

File: components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt [hunk 2]
```diff
@@ -100,7 +99,15 @@ abstract class AbstractGitpodPortForwardingService : GitpodPortForwardingService
         return completableFuture
     }
 
+    private fun isLocalPortForwardingDisabled(): Boolean {
+        return System.getenv("GITPOD_DISABLE_JETBRAINS_LOCAL_PORT_FORWARDING")?.toBoolean() ?: false
+    }
+
     private fun syncPortsListWithClient(response: Status.PortsStatusResponse) {
+        if (isLocalPortForwardingDisabled()) {
```
Comment (akosyakov) [inline components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/AbstractGitpodPortForwardingService.kt:107]: it probably should go in `startForwarding` that we only disable forwarding, not exposure

File: components/server/src/workspace/workspace-starter.ts [hunk 3]
```diff
@@ -1552,16 +1552,18 @@ export class WorkspaceStarter {
         sysEnvvars.push(orgIdEnv);
 
         const client = getExperimentsClientForBackend();
-        const [isSetJavaXmx, isSetJavaProcessorCount] = await Promise.all([
+        const [isSetJavaXmx, isSetJavaProcessorCount, disableJetBrainsPortExpose] = await Promise.all([
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-starter.ts:1555]: `disableJetBrainsPortExpose` should be renamed, it is not about port exposure

**Comment by mustard-mh** [comment]
Feedback addressed, right now we only disable local ports forwarding, other ports features should work well like before, i.e.
- Show notification when ports is opened
- Able to click on 'Open in Browser' on notifications or in ports view double click URLs

There's a known issue that 'Copy Web URL' is not working ENT-863, we will address it separately

@iQQBot could you please help test again 🙏

**Pull request merged**